xla/service/gpu/ir_emitter_unnested.cc

```diff
@@ -2557,63 +2558,77 @@ static std::optional<GlobalDeviceId> DeviceConstraint(
+  TF_ASSIGN_OR_RETURN(bool is_src_host_memory, HasHostMemorySpace(shape, 1));
+  return is_dst_host_memory || is_src_host_memory;
+}
+
 absl::Status IrEmitterUnnested::EmitCopyStartThunk(
-    const HloCopyStartInstruction* instr) {
+    const HloCopyStartInstruction* copy_start_instr) {
   // copy-start has a tuple shape: {host, device, context},
   // or {device, host, context}.
   // Only the destination shape is needed to get the output buffer.
   TF_ASSIGN_OR_RETURN(BufferAllocation::Slice dst_buffer,
-                      GetAllocationSliceForHlo(instr,
+                      GetAllocationSliceForHlo(copy_start_instr,
                                                /*ShapeIndex=*/{0}));
 
-  const HloInstruction* src = instr->operand(0);
+  const HloInstruction* src = copy_start_instr->operand(0);
   const Shape& input_shape = src->shape();
   TF_ASSIGN_OR_RETURN(BufferAllocation::Slice src_buffer,
                       GetAllocationSliceForHlo(src, {}));
-  Shape shape = instr->shape();
+  Shape shape = copy_start_instr->shape();
   CHECK(shape.IsTuple());
-
-  if (shape.mutable_tuple_shapes(0)->has_layout() &&
-      shape.mutable_tuple_shapes(0)->mutable_layout()->memory_space() ==
-          static_cast<int>(stream_executor::MemoryType::kHost)) {
-    VLOG(3) << "Device to Host: host memory space "
-            << static_cast<int>(stream_executor::MemoryType::kHost);
-    auto thunk = std::make_unique<DeviceToHostCopyThunk>(
-        Thunk::ThunkInfo::WithProfileAnnotation(instr),
-        /*source_buffer=*/src_buffer,
-        /*destination_buffer=*/dst_buffer,
-        /*mem_size=*/ShapeUtil::ByteSizeOf(input_shape));
-    AddThunkToThunkSequence(std::move(thunk));
-    return absl::OkStatus();
-  }
-  if (shape.mutable_tuple_shapes(1)->has_layout() &&
-      shape.mutable_tuple_shapes(1)->mutable_layout()->memory_space() ==
-          static_cast<int>(stream_executor::MemoryType::kHost)) {
-    VLOG(3) << "Host to Device from the host memory space "
-            << static_cast<int>(stream_executor::MemoryType::kHost);
-    ;
-    auto thunk = std::make_unique<HostToDeviceCopyThunk>(
-        Thunk::ThunkInfo::WithProfileAnnotation(instr),
-        /*source_buffer=*/src_buffer,
-        /*destination_buffer=*/dst_buffer,
-        /*mem_size=*/ShapeUtil::ByteSizeOf(input_shape));
-    AddThunkToThunkSequence(std::move(thunk));
-    return absl::OkStatus();
-  }
-
-  // Disabled the generation of memcpy D2D as only H2D and D2H are useful
-  // for memory offload now.
-
-  auto thunk = std::make_unique<DeviceToDeviceCopyThunk>(
-      Thunk::ThunkInfo::WithProfileAnnotation(instr),
+  TF_ASSIGN_OR_RETURN(bool is_dst_host_memory, HasHostMemorySpace(shape, 0));
+  TF_ASSIGN_OR_RETURN(bool is_src_host_memory, HasHostMemorySpace(shape, 1));
+  if (is_dst_host_memory == is_src_host_memory) {
+    return absl::InternalError(
+        "Copy-start doesn't have correct host memory space color");
```

Comment:
Nit: maybe give a more descriptive error message, plus also add instruction to the error message?

Reply:
Sure. Add the instruction to the error message.